From 78fd32628bf6c24188cb3da347751b7ea2c1fb94 Mon Sep 17 00:00:00 2001
From: Yang Yingliang <yangyingliang@huawei.com>
Date: Thu, 21 Jan 2021 17:16:04 +0800
Subject: [PATCH 1003/2065] [Backport] mbigen: fix mbigen driver defer probe

maillist inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I56R9J

Reference: http://patchwork.huawei.com/patch/160461

-------------------------------------------------

Current mbigen driver uses module_platform_driver() to
call init function, but pl011 driver uses arch_initcall().
So pl011 driver will init earlier than mbigen driver and
pl011 will get irq failed.

Fix this problem by using arch_initcall in mbigen driver.

Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: zhangyi (F) <yi.zhang@huawei.com>
Signed-off-by: Yi Yang <yiyang13@huawei.com>
---
 drivers/irqchip/irq-mbigen.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/irqchip/irq-mbigen.c b/drivers/irqchip/irq-mbigen.c
index 8729b8a6b54d..fc05e23938cd 100644
--- a/drivers/irqchip/irq-mbigen.c
+++ b/drivers/irqchip/irq-mbigen.c
@@ -402,7 +402,18 @@ static struct platform_driver mbigen_platform_driver = {
 	.probe			= mbigen_device_probe,
 };
 
-module_platform_driver(mbigen_platform_driver);
+static int __init mbigen_init(void)
+{
+	return platform_driver_register(&mbigen_platform_driver);
+}
+
+static void __exit mbigen_exit(void)
+{
+	platform_driver_unregister(&mbigen_platform_driver);
+}
+
+arch_initcall(mbigen_init);
+module_exit(mbigen_exit);
 
 MODULE_AUTHOR("Jun Ma <majun258@huawei.com>");
 MODULE_AUTHOR("Yun Wu <wuyun.wu@huawei.com>");
-- 
2.17.1

